bftengine/src/bftengine/ControllerWithSimpleHistory.cpp

```diff
@@ -143,10 +153,14 @@ namespace bftEngine
 					const float factorForFastOptimisticPath = ControllerWithSimpleHistory_debugUpgradeFactorForFastOptimisticPath; // 0.95F;
 					const float factorForFastPath = ControllerWithSimpleHistory_debugUpgradeFactorForFastPath; //  0.85F;
 
-					if (cyclesWithFullCooperation >= (factorForFastOptimisticPath * EvaluationPeriod))
+					if (cyclesWithFullCooperation >= (factorForFastOptimisticPath * EvaluationPeriod)) {
 						currentFirstPath = CommitPath::OPTIMISTIC_FAST;
-					else if (!onlyOptimisticFast && (cyclesWithFullCooperation + cyclesWithPartialCooperation >= (factorForFastPath * EvaluationPeriod)))
+                                                metrics_first_path_.Get().Set(CommitPathToStr(currentFirstPath));
+                                        }
+					else if (!onlyOptimisticFast && (cyclesWithFullCooperation + cyclesWithPartialCooperation >= (factorForFastPath * EvaluationPeriod))) {
 						currentFirstPath = CommitPath::FAST_WITH_THRESHOLD;
+                                                metrics_first_path_.Get().Set(CommitPathToStr(currentFirstPath));
```

Comment:
The “Controllers” are objects may be passed as arguments to the replicas. We currently assume that their member functions do not have any implicit side effect. 
Is it possible to do the work outside the Controller object?

Reply:
Ok. I fixed this, and kept the overhead low, by returning a bool indicating when the firstCommitPath is changed on return from `onNewSeqNumberExecution`. When that happens the replica retrieves the firstCommitPath from the executor and sets the metric.